[(BITS-VIB)](https://github.com/BITS-VIB)
## Comparing NGS mappers
==========

Code developped for a white paper published on our Wiki.

Full information about the RNASeq hands-on training can be found on our <a href="http://wiki.bits.vib.be" target="_blank">BITS WIKI</a> in the dedicated pages starting at <a href="http://wiki.bits.vib.be/index.php/Compare_Popular_Read_Mappers" target="_blank">Compare_Popular_Read_Mappers</a>

The code posted here will be maintained and might be different from that cointained in the virtualbox image used during the training.

This repo presents two folders for

* [scripts](scripts): code relative to the unix command-line manipulation of the reads (QC, mapping, post-processing, ...)
* [R-scripts](R-scripts):  the R/bioconductor analysis of the gene counts obtained from the mapping of the reads (EdgeR, Deseq2, RobiNA, ...)

Related work can be found in many places, please find here a starter

* <a href="http://www.ecseq.com/support/benchmark">Benchmark A comparison of different NGS read mappers</a>

<h4>Please send comments and feedback to <a href="mailto:bits@vib.be">bits@vib.be</a></h4>

------------

![Creative Commons License](http://i.creativecommons.org/l/by-sa/3.0/88x31.png?raw=true)

This work is licensed under a [Creative Commons Attribution-ShareAlike 3.0 Unported License](http://creativecommons.org/licenses/by-sa/3.0/).
